File: Proc-0/src/main/kotlin/Source/TypeDescription.kt
package proc.Source.Symbol

/**
 * В языках, имеющих типы данных, проверка их совместимости является одной из важнейших задач компилятора.
 * Проверка основана на атрибуте `ObjectDescription.Type`, вносимом в каждую запись таблицы символов.
 */
class TypeDescription {

	/**
	 * Представление типа.
	 */
	public enum class TypeForm
	{
		Class,
		Array,
		Boolean,
		Integer,
	}

	/**
	 * Тип описываемого объекта.
	 */
	public var Form: TypeForm

	/**
	 * Размер элементарного типа, определяемого архитектурой?
	 */
	public var Size: Int

	/**
	 * Количество элементов массива.
	 */
	public var Length: Int

	/// <summary>
	/// Элементы класса.
	/// </summary>
	public var Fields: List<ObjectDescription>

	/// <summary>
	/// Тип элементов массива.
	/// </summary>
	public var Base: TypeDescription?

	init {
		Form = TypeForm.Class
		Size = 0
		Length = 0
		Base = null
		Fields = listOf<ObjectDescription>()
	}
}

File: Proc-0/Source/TypeDescription.cs
//namespace Source.Symbol;

/// <summary>
/// В языках, имеющих типы данных, проверка их совместимости является одной из важнейших задач компилятора.
/// Проверка основана на атрибуте <see cref="ObjectDescription.Type"/>, вносимом в каждую запись таблицы символов.
/// </summary>
class TypeDescription
{
	/// <summary>
	/// Представление типа.
	/// </summary>
	public enum TypeForm
	{
		Class,
		Array,
		Boolean,
		Integer,
	}

	/// <summary>
	/// Тип описываемого объекта.
	/// </summary>
	public TypeForm Form;

	/// <summary>
	/// Размер элементарного типа, определяемого архитектурой?
	/// </summary>
	public int Size;

	/// <summary>
	/// Количество элементов массива.
	/// </summary>
	public int Length;

	/// <summary>
	/// Элементы класса.
	/// </summary>
	public List<ObjectDescription> Fields;

	/// <summary>
	/// Тип элементов массива.
	/// </summary>
	public TypeDescription Base;
}
